src/lib/reasoners/rel_utils.ml

```diff
@@ -193,126 +193,179 @@ end = struct
     env, { Sig_rel.assume = assume_nontrivial_eqs eqs la; remove = [] }
 end
 
-module Congruence : sig
-  (** The [Congruence] module implements a simil-congruence closure algorithm on
-      semantic values.
+module type Constraint = sig
+  type t
+  (** The type of bit-vector constraints.
+
+      Constraints are associated with a justification as to why they are
+      currently valid. The justification is only used to update domains,
+      identical constraints with different justifications will otherwise behave
+      identically (and, notably, will compare equal).
+
+      Bit-vector constraints contains semantic values / term representatives of
+      type [X.r]. We maintain the invariant that the semantic values used inside
+      the constraints are *class representatives* i.e. normal forms wrt the `Uf`
+      module, i.e. constraints have a normalized representation. Use `subst` to
+      ensure normalization. *)
+
+  val pp : t Fmt.t
+  (** Pretty-printer for constraints. *)
+
+  val compare : t -> t -> int
+  (** Comparison function for constraints.
+
+      The comparison only depends on the constraint representation: two
+      constraints with identical representations but different explanations will
+      compare equal.
+
+      {b Note}: The comparison function is arbitrary and has no semantic
+      meaning. You should not depend on any of its properties, other than it
+      defines an (arbitrary) total order on constraint representations. *)
 
-      It provides an interface to register some semantic values of interest, and
-      for applying a callback when the representative of those registered values
-      change.
-  *)
+  val subst : Explanation.t -> X.r -> X.r -> t -> t
+  (** [subst ex p v cs] replaces all the instaces of [p] with [v] in the
+      constraint.
+
+      Use this to ensure that the representation is always normalized.
+
+      The explanation [ex] justifies the equality [p = v]. *)
+
+  val fold_leaves : (X.r -> 'a -> 'a) -> t -> 'a -> 'a
+
+  type domain
+
+  val propagate : t -> domain -> domain
+  (** [propagate c dom] propagates the constraints [c] in [d] and returns the
+      new domains. *)
```

Comment:
```suggestion
      new domain. *)
```

Reply:
The `domain` type is expected to contain domains for many variables; I will rename it (and add a short documentation as well).